File: server/policy-ocr-database-target.mjs
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';

function canonicalPath(value) {
  const absolutePath = path.resolve(String(value || '').trim());
  let existingPath = absolutePath;
  const missingSegments = [];
  while (!fs.existsSync(existingPath)) {
    const parentPath = path.dirname(existingPath);
    if (parentPath === existingPath) return absolutePath;
    missingSegments.unshift(path.basename(existingPath));
    existingPath = parentPath;
  }
  return path.join(fs.realpathSync(existingPath), ...missingSegments);
}

function readConfiguredDevelopmentDatabasePath(projectRoot) {
  try {
    const configPath = path.join(projectRoot, '.runtime', 'local', 'policy-ocr-env.json');
    const payload = JSON.parse(fs.readFileSync(configPath, 'utf8'));
    return String(payload?.POLICY_OCR_APP_DB_PATH || '').trim();
  } catch {
    return '';
  }
}

function databaseTargetError(code, message, details = []) {
  const error = new Error([message, ...details].filter(Boolean).join('\n'));
  error.code = code;
  return error;
}

function normalizedProfile(profile, env) {
  const value = String(profile || env.POLICY_OCR_PROFILE || '').trim().toLowerCase();
  if (value === 'prod' || value === 'production' || (!value && env.NODE_ENV === 'production')) return 'prod';
  return 'dev';
}

export function assertNotLegacyPolicyOcrDatabasePath({ projectRoot, dbPath }) {
  if (String(dbPath || '').trim() === ':memory:') return ':memory:';
  const resolvedPath = canonicalPath(dbPath);
  const legacyPath = canonicalPath(path.join(projectRoot, '.runtime', 'local', 'policy-ocr.sqlite'));
  if (resolvedPath === legacyPath) {
    throw databaseTargetError(
      'POLICY_OCR_LEGACY_DATABASE_TARGET',
      '已拒绝写入旧开发数据库。',
      [`旧库: ${resolvedPath}`],
    );
  }
  return resolvedPath;
}

export function resolvePolicyOcrWriteDatabasePath({
  projectRoot,
  profile = '',
  requestedPath = '',
  env = process.env,
  homeDir = os.homedir(),
} = {}) {
  const root = path.resolve(String(projectRoot || '').trim());
  const targetProfile = normalizedProfile(profile, env);
  const configuredDevelopmentPath = readConfiguredDevelopmentDatabasePath(root);
  const defaultDevelopmentPath = path.join(homeDir, 'OCR_insurance_ssd', '.runtime', 'local', 'policy-ocr.sqlite');
  const expectedDevelopmentPath = configuredDevelopmentPath
    || String(env.POLICY_OCR_APP_DB_PATH || '').trim()
    || defaultDevelopmentPath;

  if (targetProfile === 'prod') {
    const expectedProductionPath = String(requestedPath || env.POLICY_OCR_APP_DB_PATH || '').trim()
      || path.join(root, '.runtime', 'policy-ocr.sqlite');
    const resolvedProductionPath = assertNotLegacyPolicyOcrDatabasePath({
      projectRoot: root,
      dbPath: expectedProductionPath,
    });
    if (resolvedProductionPath === canonicalPath(expectedDevelopmentPath)) {
      throw databaseTargetError(
        'POLICY_OCR_PRODUCTION_DATABASE_MISMATCH',
        '生产发布不得写入开发 SSD 数据库。',
        [`生产目标: ${resolvedProductionPath}`],
      );
    }
    return resolvedProductionPath;
  }

  const resolvedDevelopmentPath = assertNotLegacyPolicyOcrDatabasePath({
    projectRoot: root,
    dbPath: requestedPath || expectedDevelopmentPath,
  });
  const configuredTarget = assertNotLegacyPolicyOcrDatabasePath({
    projectRoot: root,
    dbPath: expectedDevelopmentPath,
  });
  if (resolvedDevelopmentPath !== configuredTarget) {
    throw databaseTargetError(
      'POLICY_OCR_DEVELOPMENT_DATABASE_MISMATCH',
      '开发写入目标与配置的 SSD 数据库不一致。',
      [
        `配置目标: ${configuredTarget}`,
        `请求目标: ${resolvedDevelopmentPath}`,
      ],
    );
  }
  return resolvedDevelopmentPath;
}

File: scripts/policy_ocr_database_target.py
import json
import os
from pathlib import Path


def _canonical(value: str | os.PathLike[str]) -> Path:
    return Path(value).expanduser().resolve(strict=False)


def _configured_development_database(project_root: Path) -> str:
    config_path = project_root / ".runtime" / "local" / "policy-ocr-env.json"
    try:
        payload = json.loads(config_path.read_text(encoding="utf-8"))
        return str(payload.get("POLICY_OCR_APP_DB_PATH") or "").strip()
    except (FileNotFoundError, json.JSONDecodeError, OSError):
        return ""


def resolve_policy_ocr_write_database(
    project_root: str | os.PathLike[str],
    requested_path: str = "",
    profile: str = "",
) -> str:
    root = _canonical(project_root)
    legacy_path = _canonical(root / ".runtime" / "local" / "policy-ocr.sqlite")
    configured_development_path = (
        _configured_development_database(root)
        or str(os.environ.get("POLICY_OCR_APP_DB_PATH") or "").strip()
        or str(Path.home() / "OCR_insurance_ssd" / ".runtime" / "local" / "policy-ocr.sqlite")
    )
    expected_development_path = _canonical(configured_development_path)
    target_profile = str(profile or os.environ.get("POLICY_OCR_PROFILE") or "").strip().lower()
    is_production = target_profile in {"prod", "production"} or (
        not target_profile and os.environ.get("NODE_ENV") == "production"
    )

    if is_production:
        target = _canonical(
            requested_path
            or os.environ.get("POLICY_OCR_APP_DB_PATH")
            or root / ".runtime" / "policy-ocr.sqlite"
        )
        if target == legacy_path:
            raise RuntimeError(f"已拒绝写入旧开发数据库: {target}")
        if target == expected_development_path:
            raise RuntimeError(f"生产发布不得写入开发 SSD 数据库: {target}")
        return str(target)

    target = _canonical(requested_path or expected_development_path)
    if target == legacy_path:
        raise RuntimeError(f"已拒绝写入旧开发数据库: {target}")
    if target != expected_development_path:
        raise RuntimeError(
            "开发写入目标与配置的 SSD 数据库不一致: "
            f"configured={expected_development_path} requested={target}"
        )
    return str(target)
